src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs

```diff
@@ -100,18 +103,92 @@ private void OnDestroy()
 #endif
             if (_isRecording)
             {
-                KeyboardInputActionObserver.GetInstance()?.StopRecording();
-                MouseInputActionObserver.GetInstance()?.StopRecording();
-                var theVideoDirectory = _currentVideoDirectory;
-                Task.Run(() =>
+                RGBotManager.GetInstance().ShowUploadingIndicator(true);
+                HandleRecordingSaving();
+            }
+        }
+
+        /**
+         * We separate out the async saving code since OnDestroy can't be async, and we need to wait for saving
+         * before doing API calls
+         */
+        private async void HandleRecordingSaving()
+        {
+            _isRecording = false;
+            var endTime = DateTime.Now;
+            KeyboardInputActionObserver.GetInstance()?.StopRecording();
+            MouseInputActionObserver.GetInstance()?.StopRecording();
+            var gameplaySessionDataDirectory = _currentGameplaySessionDirectoryPrefix + "_data";
+            var gameplaySessionScreenshotDirectory = _currentGameplaySessionDirectoryPrefix + "_screenshots";
+            var thumbnailPath = _currentGameplaySessionDirectoryPrefix + "_thumbnail.png";
+            await Task.Run(() =>
+            {
+                // First, save the gameplay session data
+                RGDebug.LogInfo($"Zipping state recording replay to file: {gameplaySessionDataDirectory}.zip");
+                ZipFile.CreateFromDirectory(gameplaySessionDataDirectory, gameplaySessionDataDirectory + ".zip");
+                Directory.Delete(gameplaySessionDataDirectory, true);
+                RGDebug.LogInfo($"Finished zipping replay to file: {gameplaySessionDataDirectory}.zip");
+                
+                // Then save the screenshots separately
+                RGDebug.LogInfo($"Zipping screenshot recording replay to file: {gameplaySessionScreenshotDirectory}.zip");
+                ZipFile.CreateFromDirectory(gameplaySessionScreenshotDirectory, gameplaySessionScreenshotDirectory + ".zip");
+                RGDebug.LogInfo($"Finished zipping replay to file: {gameplaySessionScreenshotDirectory}.zip");
+                
+                // Finally, we also save a thumbnail, by choosing the middle file in the screenshots
+                var screenshotFiles = Directory.GetFiles(gameplaySessionScreenshotDirectory);
+                var middleFile = screenshotFiles[screenshotFiles.Length / 2]; // this gets floored automatically
```

Comment:
This seems very arbitrary and likely to get some random(possibly even blank) frame.  I would suggest taking the first frame instead.

Reply:
don't like it, but we can change later